src/3rd_party_bundle_add.c

```diff
@@ -94,6 +98,174 @@ static bool parse_options(int argc, char **argv)
+		goto close_and_exit;
+	}
+
+	/* if the SWUPD_3RD_PARTY_BIN_DIR does not exist, attempt to create it */
+	if (!sys_file_exists(bin_directory)) {
+		if (mkdir_p(bin_directory)) {
+			ret_code = SWUPD_COULDNT_CREATE_DIR;
+			goto close_and_exit;
+		}
+	}
+
+	if (!is_dir(bin_directory)) {
+		error("The path %s for 3rd-party content exists but is not a directory\n", bin_directory);
+		ret_code = SWUPD_COULDNT_CREATE_DIR;
+		goto close_and_exit;
+	}
+
+	fp = fopen(script, "w");
+	if (!fp) {
+		error("The file %s failed to be created\n", script);
+		ret_code = SWUPD_COULDNT_CREATE_FILE;
+		goto close_and_exit;
+	}
+
+	/* get the current values of the environment */
+	env_path = getenv("PATH");
+	env_ld_path = getenv("LD_LIBRARY_PATH");
+
+	/* get the path for the 3rd-party content */
+	third_party_bin_path = str_or_die("%sbin:%susr/bin:%susr/local/bin", globals.path_prefix, globals.path_prefix, globals.path_prefix);
+	third_party_ld_path = str_or_die("%susr/lib64:%susr/local/lib64", globals.path_prefix, globals.path_prefix);
+
+	fprintf(fp, "#!/bin/bash\n");
+	fprintf(fp, "\n");
+	fprintf(fp, "export PATH=%s%s%s\n", third_party_bin_path, env_path ? ":" : "", env_path ? env_path : "");
+	fprintf(fp, "export LD_LIBRARY_PATH=%s%s%s\n", third_party_ld_path, env_ld_path ? ":" : "", env_ld_path ? env_ld_path : "");
+	fprintf(fp, "\n");
+	fprintf(fp, "%s \"$@\"\n", binary);
+
+	/* set the file mode to 0755 */
+	if (chmod(script, mode) < 0) {
+		error("Unable to set permissions for file %s\n", script);
+		ret_code = SWUPD_COULDNT_WRITE_FILE;
+	}
+
+close_and_exit:
+	if (fp) {
+		fclose(fp);
+	}
+	free_string(&binary);
+	free_string(&script);
+	free_string(&bin_directory);
+	free_string(&third_party_bin_path);
+	free_string(&third_party_ld_path);
+
+	return ret_code;
+}
+
+static enum swupd_code validate_binaries(const char *filename)
+{
+	enum swupd_code ret_code = SWUPD_OK;
+	char *bin_directory = NULL;
+	char *script = NULL;
+
+	char *tmp1 = sys_dirname(globals.path_prefix);
+	char *tmp2 = sys_dirname(tmp1);
+	bin_directory = sys_path_join(tmp2, "/bin");
+	free_string(&tmp1);
+	free_string(&tmp2);
+
+	script = sys_path_join(bin_directory, sys_basename(filename));
+
+	/* if a file already exists, report the problem to the user */
+	if (sys_file_exists(script)) {
+		error("There is already a binary called %s in %s\n", sys_basename(filename), bin_directory);
+		ret_code = SWUPD_INVALID_BINARY;
+	}
+	free_string(&script);
+	free_string(&bin_directory);
+
+	return ret_code;
+}
+
+static enum swupd_code process_bundle_binaries(struct list *installed_files, const char *msg, const char *step, proc_binary_fn_t proc_binary_fn)
+{
+	enum swupd_code ret, ret_code = SWUPD_OK;
+	struct list *iter = NULL;
+	struct file *file = NULL;
+	char *filename = NULL;
+	int count = 0;
+	int number_of_files = list_len(installed_files);
+
+	info("%s", msg);
+
+	progress_next_step(step, PROGRESS_BAR);
+	for (iter = list_head(installed_files); iter; iter = iter->next) {
```

Comment:
why do you need a list_head() here?

Reply:
I don't really since this function should run immediately after `consolidate_files()` which is going to already return list_head(). I like to do that sometimes just to make sure the head of the list is what we've got, also I think the code is easier to follow when you explicitly see you are working with the head of the list. The list_head() function should return right away if you already are at the head of the list, so there is really no penalty from doing it when you are already at the head.

Reply:
> I don't really since this function should run immediately after consolidate_files() which is going to already return list_head().

Using list_head() in locations like can mask problems instead of solving it. All functions should return the head of the list.

> I like to do that sometimes just to make sure the head of the list is what we've got, also I think the code is easier to follow when you explicitly see you are working with the head of the list.

That's an unusual implementation of a list. Usually lists should point to the element you consider the begging of the list, not any element of the list. One great advantage of this implementation is that you can call functions that will iterate partially at the list. I don't see code using list_head() or something similar in python, go, glib or any other library I have worked before.

> The list_head() function should return right away if you already are at the head of the list, so there is really no penalty from doing it when you are already at the head.

It's not about performance it's about not masking problems in other areas of the code.

Reply:
I see your point, I'll remove the list_head from there then, thanks for the explanation